packages/contracts/contracts/SmartInvoiceEscrowV2.sol

```diff
@@ -0,0 +1,444 @@
+// SPDX-License-Identifier: MIT
+// solhint-disable not-rely-on-time, max-states-count
+
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "@openzeppelin/contracts/utils/Context.sol";
+import "@openzeppelin/contracts/proxy/utils/Initializable.sol";
+import "./interfaces/ISmartInvoiceV2.sol";
+import "./interfaces/IArbitrable.sol";
+import "./interfaces/IArbitrator.sol";
+import "./interfaces/IWRAPPED.sol";
+
+import "hardhat/console.sol";
+
+// splittable digital deal lockers w/ embedded arbitration tailored for guild work
+contract SmartInvoiceEscrowV2 is
+    ISmartInvoiceV2,
+    IArbitrable,
+    Initializable,
+    Context,
+    ReentrancyGuard
+{
+    using SafeERC20 for IERC20;
+
+    uint256 public constant NUM_RULING_OPTIONS = 5; // excludes options 0, 1 and 2
+    // Note that Aragon Court treats the possible outcomes as arbitrary numbers, leaving the Arbitrable (us) to define how to understand them.
+    // Some outcomes [0, 1, and 2] are reserved by Aragon Court: "missing", "leaked", and "refused", respectively.
+    // Note that Aragon Court emits the LOWEST outcome in the event of a tie.
+
+    // solhint-disable-next-line var-name-mixedcase
+    uint8[2][6] public RULINGS = [
+        [1, 1], // 0 = refused to arbitrate
+        [1, 0], // 1 = 100% to client
+        [3, 1], // 2 = 75% to client
+        [1, 1], // 3 = 50% to client
+        [1, 3], // 4 = 25% to client
+        [0, 1] // 5 = 0% to client
+    ];
+
+    uint256 public constant MAX_TERMINATION_TIME = 63113904; // 2-year limit on locker
+
+    address public wrappedNativeToken;
+
+    // implementation info
+    uint256 public invoiceId;
+    bytes32 public implementationType;
+    uint256 public implementationVersion;
```

Comment:
I'm still not understanding how implementation version is being used. It doesn't appear to effect anything outside of being available to the user (is this for the front-end to manage feature toggles?) Do we need to emit an event with this in it so we can build it into the subgraph usage?

Reply:
It is for the frontend, yes. Initially I had planned to store this on the contract and add them to the subgraph from there, but it might make more sense to add these to the LogInvoice event and read them from event params.

Reply:
Ok, I think I see it now with the `addImplementation`. I was getting turned around slightly.

Reply:
I agree that ideally this should be logged in an event. This current method feels like a waste of storage on the invoice contract side.